File: src/HARD/76. Minimum Window Substring.ts
// 04/29/2021 HARD

// https://leetcode-cn.com/problems/minimum-window-substring/

/*
Given two strings s and t, return the minimum window in s which will contain all the characters in t. If there is no such window in s that covers all characters in t, return the empty string "".

Note that If there is such a window, it is guaranteed that there will always be only one unique minimum window in s.

 

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Example 2:

Input: s = "a", t = "a"
Output: "a"
 

Constraints:

1 <= s.length, t.length <= 105
s and t consist of English letters.
 

Follow up: Could you find an algorithm that runs in O(n) time?

 */

function minWindow(s: string, t: string): string {
  if (t.length > s.length) return '';
  const targetCount: number[] = new Array(52).fill(0);
  for (let i = 0; i < t.length; i++) targetCount[getCharCode(t[i])]++;
  let left = 0, right = 0;
  let minLen = Number.MAX_SAFE_INTEGER, ans = '';
  const slideCount: number[] = new Array(52).fill(0);
  while (right < s.length) {
    slideCount[getCharCode(s[right])]++;
    right++;
    while (check()) {
      if (right - left < minLen) {
        minLen = right - left;
        ans = s.substr(left, right - left);
      }
      slideCount[getCharCode(s[left])]--;
      left++; 
    }
  }
  return ans;
  
  function check() {
    for (let i = 0; i < 52; i++) {
      if (targetCount[i] > 0 && targetCount[i] > slideCount[i]) return false;
    }
    return true;
  }
  function getCharCode(char: string) {
    if (char >= 'a') return char.charCodeAt(0) - 'a'.charCodeAt(0) + 26;
    else return char.charCodeAt(0) - 'A'.charCodeAt(0);
  }
};

File: src/HOT100/HARD/76-MinimumWindowSubstring.cpp
//
// Created by ZiqianCheng on 2021/6/29.
//

// HARD https://leetcode-cn.com/problems/minimum-window-substring

/*
 * Given two strings s and t of lengths m and n respectively,
 * return the minimum window substring of s such that every character in t (including duplicates)
 * is included in the window. If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.

A substring is a contiguous sequence of characters within the string.

 

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
Example 2:

Input: s = "a", t = "a"
Output: "a"
Explanation: The entire string s is the minimum window.
Example 3:

Input: s = "a", t = "aa"
Output: ""
Explanation: Both 'a's from t must be included in the window.
Since the largest window of s only has one 'a', return empty string.
 

Constraints:

m == s.length
n == t.length
1 <= m, n <= 105
s and t consist of uppercase and lowercase English letters.
 

Follow up: Could you find an algorithm that runs in O(m + n) time?

 */

#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
class Solution {
public:
  string minWindow(string s, string t) {
    unordered_map<char, int> count;
    int diff = 0;
    for (auto c: t) {
      count[c]++;
      diff++;
    }
    int n = s.size();
    int left = 0, right = 0;
    int len = INT_MAX;
    int start = 0;
    while (right < n) {
      if (count.count(s[right])) {
        count[s[right]]--;
        if (count[s[right]] >= 0) diff--;
      }
      right++;
      if (!diff) {
        while (!diff) {
          if (count.count(s[left])) {
            count[s[left]]++;
            if (count[s[left]] > 0) diff++;
          }
          left++;
        }
        if (right - left + 1 < len) {
          len = right - left + 1;
          start = left - 1;
        }
      }
    }
    return len == INT_MAX ? "" : s.substr(start, len);
  }
};